File: c/data_structure/singly_linked_list.c
/** Linked List
 *
 * bool add(LinkedList* L, int32_t item)
 * bool remove_by_item(LinkedList* L, int32_t item)
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>

typedef struct LinkedList {
    struct Node* head;
} LinkedList;

typedef struct Node {
    int32_t data;
    struct Node* next;
} Node;

bool add(LinkedList* L, int32_t item);
bool remove_by_item(LinkedList* L, int32_t item);

int32_t 
main(void)
{
    // init
    LinkedList* L = (LinkedList*)malloc(sizeof(LinkedList));
    if (L == NULL) {
        return -1;
    } else {
        L->head = NULL;
    }

    // add item
    add(L, 1);
    add(L, 2);
    add(L, 3);
    add(L, 4);
    add(L, 5);
    add(L, 6);
    add(L, 7);
    add(L, 8);
    add(L, 9);
    add(L, 10);

    // traversal
    for (Node* tmp = L->head; tmp != NULL; tmp = tmp->next) {
        printf("%i ", tmp->data);
    }
    printf("\n");

    remove_by_item(L, 4);
    for (Node* tmp = L->head; tmp != NULL; tmp = tmp->next) {
        printf("%i ", tmp->data);
    }
    printf("\n");

    // free
    while (L->head != NULL) {
        Node* tmp = L->head->next;
        free(L->head);
        L->head = tmp;
    }
}

/*
 * Function: add
 * -------------
 * Add item into LinkedList
 *
 *  params: LinkedList* L, int32_t item
 *  returns: bool
 */
bool 
add(LinkedList* L, int32_t item)
{
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) return false;

    newNode->data = item;
    newNode->next = NULL;

    if (L->head == NULL) {
        L->head = newNode;
    } else {
        Node* tmp = L->head;
        while (tmp->next != NULL)
            tmp = tmp->next;

        tmp->next = newNode;
    }
    
    return true;
}

/*
 * Function: remove_by_item
 * -------------
 * Remove item in LinkedList
 *
 *  params: LinkedList* L, int32_t item
 *  returns: bool
 */
bool 
remove_by_item(LinkedList* L, int32_t item)
{
    if (L->head == NULL) {
        return false;
    }

    if (L->head->data == item) {
        L->head = L->head->next;
    } else {
        Node* before = NULL;
        Node* curr = L->head;

        while (curr->data != item) {
            if (curr->next == NULL) {
                printf("item is not exists.\n");
                return false;
            }
            before = curr;
            curr = curr->next;
        }
        before->next = curr->next;
        free(curr);
    }
    
    return true;
}

//#include <stdio.h>
//#include <stdlib.h>
//#include <stdbool.h>
//
//typedef struct Node
//{
//    int32_t data;
//    struct Node* next;
//} Node;
//
//bool add(Node** head, int32_t item);
//
//int32_t main(void)
//{
//    // init
//    Node* head = NULL;
//
//    // add item
//    add(&head, 1);        // head를 NULL로 초기화 했기 때문에 포인터의 주소를 쏨.
//    add(&head, 2);        // if malloc() 함수로 초기화를 하면? head->data가 할당된 메모리에 전부 0으로 들어감.
//    add(&head, 3);        // 그래서 안됨. 출력이 0 1 2 3 4 5 가 됨.
//    add(&head, 4);
//    add(&head, 5);
//
//    // traversal
//    for (Node* tmp = head; tmp != NULL; tmp = tmp->next)
//        printf("%i ", tmp->data);
//
//    // free
//    while (head != NULL)
//    {
//        Node* tmp = head->next;
//        free(head);
//        head = tmp;
//    }
//}
//
//bool add(Node** head, int32_t item)
//{
//    Node* newNode = (Node*)malloc(sizeof(Node));
//    if (newNode == NULL) return false;
//
//    newNode->data = item;
//    newNode->next = NULL;
//
//    if (*head == NULL)
//        *head = newNode;
//    else
//    {
//        Node* tmp = *head;
//        while (tmp->next != NULL)
//            tmp = tmp->next;
//        tmp->next = newNode;
//    }
//    
//    return true;
//}

File: java/dataStructure/SinglyLinkedList.java
/**
 * SinglyLinkedList
 */

package dataStructure;

import java.util.Iterator;

public class SinglyLinkedList<T> {
    private Node<T> head;
    private int size = 0;

    private static class Node<T> {
        private T data;
        private Node<T> next;

        Node(T item) {
            this.data = item;
        }

        public String toString() {
            return "data: " + this.data;
        }
    }

    /**
     * Appends item to the end of this list.
     */
    public void add(T item) {
        addLast(item);
    }

    /**
     * Add into idx
     */
    public void add(int idx, T item) {
        if (idx < 0 || idx > size) {
            throw new IndexOutOfBoundsException("" + idx);
        }

        if (idx == 0) {
            addFirst(item);
        }

        Node<T> before = node(idx-1);
        addAfter(before, item);
    }

    /**
     * Insert into First
     */
    public void addFirst(T item) {
        Node<T> newNode = new Node<>(item);

        newNode.next = head;
        head = newNode;
        size++;
    }

    /**
     * Insert into Last
     */
    public void addLast(T item) {
        Node<T> newNode = new Node<>(item);

        if (head == null) {
            head = newNode;
        } else {
            Node<T> curr = head;
            while (curr.next != null) {
                curr = curr.next;
            }
            curr.next = newNode;
        }
        size++;
    }

    /**
     * Insert into before.next
     */
    private void addAfter(Node<T> before, T item) {
        Node<T> newNode = new Node<>(item);

        newNode.next = before.next;
        before.next = newNode;
        size++;
    }

    /**
     * Remove First Node<T>
     */
    public T remove() {
        return removeFirst();
    }

    public T remove(int idx) {
        if (idx < 0 || idx >= size) {
            throw new IndexOutOfBoundsException("" + idx);
        }

        if (idx == 0) {
            return removeFirst();
        }

        Node<T> before = node(idx-1);
        return removeAfter(before);
    }

    /**
     * Remove By Item
     */
    public boolean removeByItem(T item) {
        int idx = indexOf(item);

        if (idx != -1) {
            remove(idx);
            return true;
        }
        return false;
    }

    /**
     * Remove First 
     */
    public T removeFirst() {
        if (head == null)
            return null;

        T tmp = head.data;
        head = head.next;
        size--;

        System.out.println("first node is removed.");
        return tmp;
    }

    private T removeAfter(Node<T> before) {
        if (before.next == null) {
            return null;
        }
        
        T tmp = before.next.data;
        before.next = before.next.next;
        size--;
        return tmp;
    }

    private Node<T> node(int idx) {
        if (idx < 0 || idx >= size) {
            return null;
        }
        
        Node<T> curr = head;
        for (int i=0; i<idx; i++) {
            curr = curr.next;
        }
        
        return curr;
    }

    public int indexOf(T item) {
        Node<T> curr = head;
        int idx = 0;
        while (curr != null) {
            if (curr.data == item) {
                return idx;
            }
            curr = curr.next;
            idx++;
        }
        System.out.println("data " + item + " is not in linkedlist");
        return -1;
    }

    public int size() {
        return this.size;
    }

    public Iterator<T> iterator() {
        return new MyIterator();
    }

    private class MyIterator implements Iterator<T> {
        private Node<T> nextNode;

        public MyIterator() {
            this.nextNode = head;
        }

        public boolean hasNext() {
            return this.nextNode != null;
        }

        public T next() {
            if (nextNode == null) {
                throw new java.util.NoSuchElementException();
            }

            T tmp = nextNode.data;
            nextNode = nextNode.next;
            return tmp;
        }
    }

    public static void main(String[] args) {
        // SinglyLinkedList<Integer> a = new SinglyLinkedList<>();
        // System.out.println(a.head);                     // prints null
        // System.out.println();

        // a.remove(10);                                   // prints "It's empty.""
        // System.out.println();

        // a.addLast(10);
        // System.out.println(a.head);                     // prints "data: 10"
        // System.out.println(a.head.next);                // prints null
        // System.out.println();

        // a.addLast(20);
        // System.out.println(a.head);                     // prints "data: 10"
        // System.out.println(a.head.next);                // prints "data: 20"
        // System.out.println();

        // a.addLast(30);
        // System.out.println(a.head);                     // prints "data: 10"
        // System.out.println(a.head.next);                // prints "data: 20"
        // System.out.println(a.head.next.next);           // prints "data: 30"
        // System.out.println();

        // a.addFirst(5);
        // System.out.println(a.head);                     // prints "data: 5"
        // System.out.println(a.head.next);                // prints "data: 10"
        // System.out.println(a.head.next.next);           // prints "data: 20"
        // System.out.println(a.head.next.next.next);      // prints "data: 30"
        // System.out.println();

        // a.remove(20);                                   // prints "data 20 삭제 완료.""
        // System.out.println(a.head);                     // prints "data: 5"
        // System.out.println(a.head.next);                // prints "data: 10"
        // System.out.println(a.head.next.next);           // prints "data: 30"
        // System.out.println();

        // a.remove(20);                                   // prints "data 20 is not in linkedlist"

        //SinglyLinkedList<Integer> a = new SinglyLinkedList<>();
        //a.add(9);
        //a.add(8);
        //a.add(7);
        //a.add(6);
        //a.add(5);
        //a.add(4);
        //a.add(3);
        //a.add(2);
        //a.add(1);
        //a.add(0);
        //System.out.println(a.size);

        //a.removeByItem(9);
        //System.out.println(a.size);

        //a.remove(9);
        //System.out.println(a.size);

        //for (int k=0; k<a.size; k++) {
            //System.out.print(a.node(k).data + " ");
        //}
        //System.out.println();

        SinglyLinkedList<String> a = new SinglyLinkedList<>();
        a.add("a");
        a.add("b");
        a.add("c");
        a.add("d");
        System.out.println(a.indexOf("c"));

        Iterator<String> iter = a.iterator();
        while (iter.hasNext())
            System.out.print(iter.next() + " ");
        System.out.println();
    }
}
